### ISSUE TITLE
Set ngClass/ngStyle from props

### ISSUE THREAD

**Issue by spaceribs** [storybookjs/storybook#12946]
Apologies, I'm still a bit of a noob writing stories with Storybook. When building out angular component stories, I'd like to be able to set a class to style the component in different ways. Currently I'm using templates to perform that:

```ts
export const Flashing = args => ({
  moduleMetadata: {
    imports: [DotModule],
  },
  template: `
  <app-dot
    [color]="color"
    class="${DotState.Flashing}"
  ></app-dot>
  `,
  props: {
    color: '#00F',
    ...args,
  },
});
```

I'd actually much prefer the following format, but it doesn't seem to work:

```ts
export const Flashing = args => ({
  moduleMetadata: {
    imports: [DotModule],
  },
  component: DotComponent,
  props: {
    color: '#00F',
    ngClass: ['flashing'],
    ...args,
  },
});
```

I'd be happy to help if this makes any sense to implement or there are no proposed alternatives.

**Comment by Marklb** [comment]
I have been thinking about this and my main hesitation is adding custom implementations for directives, because if the functionality isn't exactly the same or someone has another directive/component that thinks it can inject a reference to the directive, it may be confusing.

If we could make directives work without a `template` story, then this would be possible, by just using the directive. I don't know how well it would work yet, but I wonder if we could do something like I am trying to do in #12438 in the core renderer. That way we could let Angular handle the rendering from a template and be able to support directives.

I haven't seen any dynamic Component libraries support directives, without limitations. It makes sense to not build a template string for rendering a component in an app, but for Storybook's use case, it may make sense to do that.

**Comment by spaceribs** [comment]
Makes sense, I definitely feel as though the `component` param is an oversimplification of what angular considers to be a special directive, from [the docs](https://angular.io/guide/architecture-components#directives):

> A component is technically a directive. However, components are so distinctive and central to Angular applications that Angular defines the @Component() decorator, which extends the @Directive() decorator with template-oriented features.

We could reconceptualize this as a collection:

```
directives: [
    {
        directive: DotComponent,
        props: { color: '#00F' }
    },
    {
        directive: NgClass,
        props: { ngClass: ['flashing'] }
    }
]
```

I don't love it, but it's certainly closer to how Angular works architecturally.

**Comment by Marklb** [comment]
@spaceribs As far as how they would be defined, something like that is what I was thinking.

The blocker that I am thinking we would need, to dynamically create the story component from a template in the renderer, is explained in this issue: https://github.com/angular/angular/issues/8785. _My internet is down and it's difficult to search on my phone, but I will link the right issue tomorrow if that isn't the one I was thinking of._ We would basically just be composing the directives and there just doesn't seem to be an easy way to do it from js still, for some reason.

**Comment by spaceribs** [comment]
@ThibaudAV @Marklb per: https://github.com/storybookjs/storybook/pull/13215#issuecomment-732729849 can we close this or does it still need to be documented?

**Comment by ThibaudAV** [comment]
No, I don't think so :/ . What has been done in https://github.com/storybookjs/storybook/pull/13215#issuecomment-732729849 does not correct this problem. 
IMHO: 2 functionalities are opposed because we want to do everything in a single `props` object, which makes it impossible to support (actually) this issue.

Currently props properties can be of these types: 
- angular input / output
- internal component properties (property that is not an input/output of current component)

<details><summary>Current behaviour : (simple explanation)</summary>

This component -> 
```js
@Component({selector: 'my-component', template: '' }) 
class MyComponent {

@Input() input: string;

@Output() output = new EventEmitter<string>();

localProp = 'only for template'
}
```

With this story -> 
```js
export const Default = args => ({
  component: MyComponent,
  props: {
    input: 'new input value',
    output: action('output'),
    localProp: 'new localProp'
  },
});
```
Generates this template: `<my-component [input]="input" (output)="output($event)"></my-component>` which is added to internal angular component : `storybook-wrapper` 

The `storybook-wrapper` component contains all props values as properties, which allows it to pass them on to components present in its template (as a real use of this one in angular). For props remaining that are not input/output (like `localProp`) `storybook-wrapper` retrieves the instance of the child component and modifies it directly. 

</details>

It should be possible to remove local properties of props in order to consider all default props as a template attribute.
and thus be able to add `class`, `ngClass`, `style`,  other directive and even why not an *ngIf
Local props could go into another `localProps` (`componentInternalProps` 🤷‍♂️ ) object
I think it would be more logical and intuitive. 

But this causes braking changes. 

There may be other possibilities. And even this one requires some feedback before doing a PR. i think 🤔

**Comment by spaceribs** [comment]
damn @ThibaudAV you totally got my hopes up with:
> With this solution it works without any problems. If my quick test is correct
let me know of course if I can help plan out this functionality within the grander scheme of things @Marklb

**Comment by ThibaudAV** [comment]
😄  haha sorry. I didn't have all the subtlety in mind yet. 
I've got some 2 pr's in progress here. I am waiting for feedback and also the end of holidays to finish them. Maybe after 🤷‍♂️

**Comment by ndelangen** [comment]
Is this still an issue in 7.0 beta @spaceribs @ThibaudAV ?

**Comment by Marklb** [comment]
@ndelangen Yes. There is no way to identify a property in `props` as a directive. There is a chance the requested syntax would currently work, but the directive would most likely only receive the initial value and not see additional props changes.

 I have a fairly large change I want to make to the Angular renderer, which I hope to propose in a PR soon, that would potentially make supporting this a little easier. The change I will be proposing reads more information about the compiled template and lets Angular handle more property updating from the template, instead of all properties being directly updated on the component instance.

**Comment by ndelangen** [comment]
You should probably talk to @kroeder and @valentinpalkovic about this PR you have in mind! 👍

**Comment by valentinpalkovic** [comment]
@Marklb Let me know, if you need any support.

**Comment by valentinpalkovic** [comment]
Thanks for the question. This was on an old `@storybook/angular` (Vue 2-era timeframe); Angular minimum is now 18+ in Storybook 9 and the story authoring API (CSF3 with `args`/`argTypes`) is the canonical way to bind props. `ngClass`/`ngStyle` should be configured via wrapper components or `applicationConfig`. Closing as stale.
